File: rust/examples/linux/comms/rfm9x/demo/src/main.rs
/* RFM95W demo — two-node round-trip link test.
 *
 * Hardware: two RFM95W modules wired back-to-back (or two boards running
 * the same code). Both configured for 868 MHz, SF=7, BW=125 kHz, CR 4/5.
 *
 * Runs 10 TX/RX iterations: transmits an incrementing 4-byte counter, then
 * immediately waits up to 1 s for the peer to echo it back. Prints the
 * round-trip time and per-packet RSSI/SNR on success, then reports the
 * total packet loss.
 */
use embedded_hal_bus::spi::ExclusiveDevice;
use linux_embedded_hal::SpidevBus;
use periph::chips::comms::Rfm95Full;
use spidev::{SpiModeFlags, Spidev, SpidevOptions};

struct NullCs;
impl embedded_hal::digital::ErrorType for NullCs {
    type Error = core::convert::Infallible;
}
impl embedded_hal::digital::OutputPin for NullCs {
    fn set_low(&mut self) -> Result<(), Self::Error> { Ok(()) }
    fn set_high(&mut self) -> Result<(), Self::Error> { Ok(()) }
}

fn main() {
    let dev_path = std::env::var("SPI_DEV").unwrap_or_else(|_| "/dev/spidev0.0".to_string());
    let mut spi = Spidev::open(dev_path).expect("open spi");
    spi.configure(&SpidevOptions::new()
        .max_speed_hz(5_000_000)
        .mode(SpiModeFlags::SPI_MODE_0)
        .build()).expect("configure spi");
    let bus = SpidevBus(spi);
    let device = ExclusiveDevice::new_no_delay(bus, NullCs).expect("spi device");

    let mut radio = Rfm95Full::new(device, 868_000_000).expect("init RFM95");    // Create RFM95W driver, (spi, frequency_hz=868e6) → Result

    // --- Configure for short-range link test ---
    // SF7 / 125 kHz / 4/5 keeps airtime low so the round-trip fits in a 1 s window;
    // +17 dBm on PA_BOOST gives enough link margin for desk-top loop-back.
    radio.inner.inner.configure(7, 125.0, 5).expect("configure");                  // Configure LoRa modem, (sf=7, bandwidth_khz=125.0, coding_rate=5) → Result<()>
    radio.inner.inner.set_tx_power(17, true).expect("tx_power");                   // Set TX power, (power_dbm=17, use_pa_boost=true) → Result<()>

    let mut loss = 0u32;
    let total = 10;
    for n in 0..total {
        let tx = (n as u32).to_be_bytes();
        let t0 = std::time::SystemTime::now();
        radio.inner.inner.send(&tx).expect("send");                                // Send packet, (data=&[u8] ≤255 B) → Result<()>

        let rx = radio.inner.inner.receive(1000).expect("receive");                // Receive single packet, (timeout_ms=1000) → Result<Option<[u8;256]>>
        let t1 = std::time::SystemTime::now();

        let rtt = t1.duration_since(t0).map(|d| d.as_millis()).unwrap_or(0);
        if let Some(buf) = rx.as_ref() {
            if &buf[..4] == tx {
                let rssi = radio.inner.inner.last_packet_rssi().unwrap_or(0.0);   // Last packet RSSI, () → Result<f32> dBm
                let snr  = radio.inner.inner.last_packet_snr().unwrap_or(0.0);    // Last packet SNR, () → Result<f32> dB
                println!("[{}] echo rtt={} ms  rssi={:.1}  snr={:.1}", n, rtt, rssi, snr);
            } else {
                loss += 1;
                println!("[{}] no echo", n);
            }
        } else {
            loss += 1;
            println!("[{}] no echo", n);
        }
        std::thread::sleep(std::time::Duration::from_millis(200));
    }
    println!("done — {}/{} successful, {} lost", total - loss, total, loss);
}

File: cpp/examples/linux/comms/RFM9x/demo/main.cpp
/* RFM95W demo — two-node round-trip link test.
 *
 * Hardware: two RFM95W modules, one running this program and one running a
 * receive-and-echo loop. Both are configured for 868 MHz, SF=7, BW=125 kHz,
 * CR 4/5 (use RFM96Full at 433 MHz for the low-band RFM96/98 modules).
 *
 * Runs 10 TX/RX iterations: transmits an incrementing 4-byte big-endian
 * counter, then waits up to 1 s for the peer to echo it back. Prints the
 * round-trip time and per-packet RSSI/SNR on success, then reports the
 * total packet loss.
 */
#include <cstdio>
#include <ctime>
#include <unistd.h>
#include "SPIConnectionLinux.h"
#include "RFM9x.h"

#ifndef TEST_SPI_BUS
#define TEST_SPI_BUS 0
#endif
#ifndef TEST_SPI_DEVICE
#define TEST_SPI_DEVICE 0
#endif

static long long now_ms() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (long long)ts.tv_sec * 1000LL + ts.tv_nsec / 1000000L;
}

int main() {
    SPIConnectionLinux connection(TEST_SPI_BUS, TEST_SPI_DEVICE, 0, 5000000); // Create SPI connection, (bus=0, device=0, mode=0, max_speed_hz=5e6 Hz) → SPIConnectionLinux
    RFM95Full radio(connection, 868000000);              // Create RFM95W driver, (connection, frequency_hz=868e6 Hz) → RFM95Full

    // --- Configure for a short-range link test ---
    // SF7 / 125 kHz / 4/5 keeps airtime low (about 40 ms for 4 bytes) so the
    // round trip fits in a 1 s window; +17 dBm on PA_BOOST gives plenty of
    // link margin for two modules on the same desk.
    radio.configure(7, 125.0f, 5);                       // Configure LoRa modem, (sf=7, bandwidth_khz=125.0 kHz, coding_rate=5) → void
    radio.set_tx_power(17, true);                        // Set TX power, (power_dbm=17 dBm, use_pa_boost=true) → void

    // --- Ping-pong loop ---
    // Each iteration sends the counter and immediately listens for the echo.
    // A missing or corrupted echo counts as a lost packet; the RSSI and SNR of
    // good echoes show the link quality in the return direction.
    const unsigned total = 10;
    unsigned loss = 0;
    for (unsigned n = 0; n < total; n++) {
        uint8_t tx[4] = { (uint8_t)(n >> 24), (uint8_t)(n >> 16), (uint8_t)(n >> 8), (uint8_t)n };

        long long t0 = now_ms();
        radio.send(tx, 4);                               // Send packet, (data, len=4) → void

        uint8_t rx[255];
        size_t got = 0;
        bool ok = radio.receive(rx, got, 1000);          // Receive single packet, (buf, len, timeout_ms=1000 ms) → bool
        long long t1 = now_ms();

        if (ok && got == 4 && rx[0] == tx[0] && rx[1] == tx[1] && rx[2] == tx[2] && rx[3] == tx[3]) {
            float rssi = radio.last_packet_rssi();       // Last packet RSSI, () → float dBm
            float snr = radio.last_packet_snr();         // Last packet SNR, () → float dB
            printf("[%u] echo rtt=%lld ms  rssi=%.1f dBm  snr=%.1f dB\n", n, t1 - t0, rssi, snr);
        } else {
            loss++;
            printf("[%u] no echo\n", n);
        }
        usleep(200 * 1000);
    }

    // --- Summary ---
    // Packet loss over the run is the headline number for link reliability.
    printf("done: %u/%u successful, %u lost\n", total - loss, total, loss);
    return 0;
}
